<a href="https://colab.research.google.com/github/Zaryn-Ooi/Malaysia-Pay-Gap-/blob/main/MalaysianPayGap_NLP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Do Malaysians like their job ?
Aim: To identify whether most malaysians are satisfied with their work or not with the use of NLP.

Data Source: Malaysian Pay Gap Instagram Account

Data Content: Text within the images of a post


In [1]:
# Import packages
import pandas as pd
import numpy as np


In [3]:
# Load data
from google.colab import files
uploaded = files.upload()

Saving posts_text.csv to posts_text.csv


In [4]:
df = pd.read_csv('posts_text.csv')
df.head()

hashtags  \
0                  ['#malaysianpaygap']   
1                  ['#malaysianpaygap']   
2                  ['#malaysianpaygap']   
3  ['#malaysianpaygap', '#mpgcreative']   
4  ['#malaysianpaygap', '#mpgcreative']   

                                            captions  likes  comments  \
0                                                NaN    205        13   
1                                                NaN    154         2   
2  Congratulations on your first job out of uni! ...    149        14   
3  Passion-driven career often met with not-so-pa...    674        20   
4  Passion-driven career often met with not-so-pa...    674        20   

     image_ids                                        image_paths  \
0  CaTj7N9P64r  /Users/aishapeng/Python/MalaysianPayGap-master...   
1  CaTiOpXvF3A  /Users/aishapeng/Python/MalaysianPayGap-master...   
2  CaUQxjsvuZk  /Users/aishapeng/Python/MalaysianPayGap-master...   
3  CaY3AMMPCvC  /Users/aishapeng/Python/MalaysianPayGap-master...   
4  CaY3AMMPCvC  /Users/aishapeng/Python/MalaysianPayGap-master...   

                                          image_text  Unnamed: 7  
0  Senior Sales Engineer in MNC\ncompany\n\n30\n\...         610  
1  Marketing & PR\n27\n\nFemale\nChinese\n\n4 yea...         323  
2  Social media executive\n26\n\nFemale\n\nMalay\...         401  
3  Animator\n\nin game/ tv series industry\n26 ch...         158  
4  Started out with RM 2000 as a\nfresh grad.\n\n...         569

In [5]:
len(df)

218

In [7]:
# Detect and remove NaN values 
df.isnull().sum()

hashtags        0
captions       20
likes           0
comments        0
image_ids       0
image_paths     0
image_text      0
Unnamed: 7      0
dtype: int64

We're only looking if there is null values in the image_text. Since there is no null values in the image_text column, we do not need to remove anything. 

In [8]:
# Detect duplicated rows
df.duplicated().sum()

0

In [9]:
df.columns

Index(['hashtags', 'captions', 'likes', 'comments', 'image_ids', 'image_paths',
       'image_text', 'Unnamed: 7'],
      dtype='object')

## Data Cleaning

In [10]:
# Find the number of unique image id
num_of_image = len(df['image_ids'].unique())
num_of_image

156

In [11]:
# Concatenate rows with same image id, and make image id as index
df = df.groupby('image_ids',as_index=True).agg(lambda x: x.tolist())

In [12]:
len(df)

156

We realised that there are 218 rows of data, however there are only 156 unique image id. This means that some of the posts(image id) have multiple images. Hence, we combined rows with the same image id. Now, there is no more duplicated image id. 

In [13]:
# Combine multiple lists into one list
df['image_text'] = df['image_text'].apply(lambda image_text: " ".join(image_text))

In [14]:
# Remove every last word of the image_text 
df['image_text']= df['image_text'].apply(lambda image_text: " ".join(image_text.split()[:-2]))

## Named Entity Recognition (NER)


In [ ]:
#import spacy
#nlp = spacy.load('en_core_web_sm')

In [ ]:
# Write a function to display basic entity info:
# def show_ents(df):
#    if df.ents:  # if it has named entities
#        for ent in df.ents:
#            print(ent.text+' - '+ent.label_+' - '+str(spacy.explain(ent.label_)))
#    else:
#        print('No named entities found.')

## Sentiment Analysis 

In [ ]:
import nltk
nltk.download('vader_lexicon')

[nltk_data] Downloading package vader_lexicon to /root/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


True

In [ ]:
# Import SentimentIntensityAnalyzer and create an sid object
from nltk.sentiment.vader import SentimentIntensityAnalyzer

sid = SentimentIntensityAnalyzer()

In [ ]:
# Use sid to append a comp_score to the dataset
df['scores'] = df['image_text'].apply(lambda image_text:sid.polarity_scores(image_text))

df['compound'] = df['scores'].apply(lambda score_dict: score_dict['compound'])

df['comp_score'] = df['compound'].apply(lambda c: 'satisfied' if c>= 0.5 else 'neutral' if c>=0.1 else 'dissatisfied')

df.head()

hashtags  \
image_ids                                                        
Ca0ueFAPokI  [['#malaysianpaygap', '#mpghealthcare'], ['#ma...   
CaRp-1uPh8l                             [['#malaysianpaygap']]   
CaT5MguPpDI                             [['#malaysianpaygap']]   
CaT6d2Yve5X                             [['#malaysianpaygap']]   
CaTg8Ijvcsw                             [['#malaysianpaygap']]   

                                                      captions         likes  \
image_ids                                                                      
Ca0ueFAPokI  [Our society is filled with so many different ...  [1547, 1547]   
CaRp-1uPh8l                                              [nan]        [1109]   
CaT5MguPpDI  [Earning only stable currencies is on everyone...         [244]   
CaT6d2Yve5X  [The many forms of wage theft in local work sc...         [327]   
CaTg8Ijvcsw                                              [nan]         [191]   

             comments                                        image_paths  \
image_ids                                                                  
Ca0ueFAPokI  [17, 17]  [/Users/aishapeng/Python/MalaysianPayGap-maste...   
CaRp-1uPh8l       [2]  [/Users/aishapeng/Python/MalaysianPayGap-maste...   
CaT5MguPpDI      [18]  [/Users/aishapeng/Python/MalaysianPayGap-maste...   
CaT6d2Yve5X      [10]  [/Users/aishapeng/Python/MalaysianPayGap-maste...   
CaTg8Ijvcsw       [4]  [/Users/aishapeng/Python/MalaysianPayGap-maste...   

                                                    image_text  \
image_ids                                                        
Ca0ueFAPokI  Behaviour Therapist (ABA Therapy) Workng w chi...   
CaRp-1uPh8l  MALAYSIA'S MEDIAN SALARY LW 19 bY a bY BELOW T...   
CaT5MguPpDI  Director 32 Male Chinese 9 years Bachelors 5-1...   
CaT6d2Yve5X  IT (Healthcare / Cybersecurity) 45 Malay RM8k ...   
CaTg8Ijvcsw  PR/Marketing in NGO sector 28 Female Malay 6 y...   

                                                        scores  compound  \
image_ids                                                                  
Ca0ueFAPokI  {'neg': 0.061, 'neu': 0.795, 'pos': 0.144, 'co...    0.9304   
CaRp-1uPh8l  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...    0.0000   
CaT5MguPpDI  {'neg': 0.071, 'neu': 0.78, 'pos': 0.149, 'com...    0.9168   
CaT6d2Yve5X  {'neg': 0.064, 'neu': 0.879, 'pos': 0.057, 'co...    0.0516   
CaTg8Ijvcsw  {'neg': 0.053, 'neu': 0.806, 'pos': 0.141, 'co...    0.7842   

               comp_score  
image_ids                  
Ca0ueFAPokI     satisfied  
CaRp-1uPh8l  dissatisfied  
CaT5MguPpDI     satisfied  
CaT6d2Yve5X  dissatisfied  
CaTg8Ijvcsw     satisfied

In [ ]:
df['comp_score'].value_counts()

satisfied       81
dissatisfied    58
neutral         17
Name: comp_score, dtype: int64

In [ ]:
# display percentage 
df['comp_score'].value_counts(normalize = True)

satisfied       0.519231
dissatisfied    0.371795
neutral         0.108974
Name: comp_score, dtype: float64

Based on the sentiment analysis , it shows that 50% of Malaysians are satisfied with their work. However, we need to further explore to see if this is true. 

In [ ]:
# display how satisfied jobs looks like
dissatisfied = df.loc[df['comp_score'] == 'dissatisfied']
dissatisfied.head()

hashtags  \
image_ids                             
CaRp-1uPh8l  [['#malaysianpaygap']]   
CaT6d2Yve5X  [['#malaysianpaygap']]   
CaTh3SgvnI5  [['#malaysianpaygap']]   
CaThAfwP7Nc  [['#malaysianpaygap']]   
CaThuNHvOiu  [['#malaysianpaygap']]   

                                                      captions   likes  \
image_ids                                                                
CaRp-1uPh8l                                              [nan]  [1109]   
CaT6d2Yve5X  [The many forms of wage theft in local work sc...   [327]   
CaTh3SgvnI5                                              [nan]   [118]   
CaThAfwP7Nc                                              [nan]   [240]   
CaThuNHvOiu                                              [nan]    [61]   

            comments                                        image_paths  \
image_ids                                                                 
CaRp-1uPh8l      [2]  [/Users/aishapeng/Python/MalaysianPayGap-maste...   
CaT6d2Yve5X     [10]  [/Users/aishapeng/Python/MalaysianPayGap-maste...   
CaTh3SgvnI5     [17]  [/Users/aishapeng/Python/MalaysianPayGap-maste...   
CaThAfwP7Nc     [20]  [/Users/aishapeng/Python/MalaysianPayGap-maste...   
CaThuNHvOiu      [2]  [/Users/aishapeng/Python/MalaysianPayGap-maste...   

                                                    image_text  \
image_ids                                                        
CaRp-1uPh8l  MALAYSIA'S MEDIAN SALARY LW 19 bY a bY BELOW T...   
CaT6d2Yve5X  IT (Healthcare / Cybersecurity) 45 Malay RM8k ...   
CaTh3SgvnI5  Project Engineer in Food manufacturing Industr...   
CaThAfwP7Nc  Senior Barista in F&B 25 2 years experience 2....   
CaThuNHvOiu  Senior Finance Logistics 28 Indian 5 years Rm4...   

                                                        scores  compound  \
image_ids                                                                  
CaRp-1uPh8l  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...    0.0000   
CaT6d2Yve5X  {'neg': 0.064, 'neu': 0.879, 'pos': 0.057, 'co...    0.0516   
CaTh3SgvnI5  {'neg': 0.148, 'neu': 0.852, 'pos': 0.0, 'comp...   -0.7736   
CaThAfwP7Nc  {'neg': 0.134, 'neu': 0.783, 'pos': 0.083, 'co...   -0.2075   
CaThuNHvOiu  {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...    0.0000   

               comp_score  
image_ids                  
CaRp-1uPh8l  dissatisfied  
CaT6d2Yve5X  dissatisfied  
CaTh3SgvnI5  dissatisfied  
CaThAfwP7Nc  dissatisfied  
CaThuNHvOiu  dissatisfied

In [ ]:
neutral = df[df['comp_score'] == 'neutral']
# neutral

## Latent Semantic Analysis (LSA)

### Stemming

In [15]:
# perform stemming to reduce words to their root
from nltk.stem.snowball import SnowballStemmer

# The Snowball Stemmer requires that you pass a language parameter
s_stemmer = SnowballStemmer(language='english')

In [16]:
def stemmed_text(imagetext):
    return ' '.join([s_stemmer.stem(word) for word in imagetext.split()])

df['image_text_lem'] = df['image_text'].apply(stemmed_text)
print(df['image_text_lem'][:5]) 
df['image_text'][:5]

image_ids
Ca0ueFAPokI    behaviour therapist (aba therapy) workng w chi...
CaRp-1uPh8l    malaysia median salari lw 19 by a by below the...
CaT5MguPpDI    director 32 male chines 9 year bachelor 5-10k ...
CaT6d2Yve5X    it (healthcar / cybersecurity) 45 malay rm 60,...
CaTg8Ijvcsw    pr/market in ngo sector 28 femal malay 6 year ...
Name: image_text_lem, dtype: object


image_ids
Ca0ueFAPokI    Behaviour Therapist (ABA Therapy) Workng w chi...
CaRp-1uPh8l    MALAYSIA'S MEDIAN SALARY LW 19 bY a bY BELOW T...
CaT5MguPpDI    Director 32 Male Chinese 9 years Bachelors 5-1...
CaT6d2Yve5X    IT (Healthcare / Cybersecurity) 45 Malay RM 60...
CaTg8Ijvcsw    PR/Marketing in NGO sector 28 Female Malay 6 y...
Name: image_text, dtype: object

In [ ]:
# for imagetext in df['image_text'].values:
#   res = ""  
#   for word in imagetext.split():
#     res += s_stemmer.stem(word)
#     print(word+' ---> '+s_stemmer.stem(word))

### Document Term Matrix

In [30]:
# count term frequency and converts each document into a vector
from sklearn.feature_extraction.text import CountVectorizer

CV = CountVectorizer(max_df = 0.20, min_df = 2, stop_words='english') # remove stop words
bag_of_words = CV.fit_transform(df['image_text_lem']) 

In [31]:
# Transform the original text message to vector
vector = CV.transform(df['image_text_lem'])

In [32]:
vector.shape

(156, 1022)

min_df reduce the words from 3075 characters to 1085.
stemming has reduced the words from 1085 to 1062.
Remove last words and reduced max_df to 0.20 have reduced 1062 to 1022..

### Singular Value Decomposition (SVD)

In [33]:
bag_of_words

<156x1022 sparse matrix of type '<class 'numpy.int64'>'
	with 5103 stored elements in Compressed Sparse Row format>

In [34]:
from sklearn.decomposition import TruncatedSVD
svd = TruncatedSVD(n_components = 2)
lsa = svd.fit_transform(bag_of_words)

In [35]:
# Constructing DataFrame from lsa
topic_encoded = pd.DataFrame(lsa, columns = ["topic_1", "topic_2"])
topic_encoded['text'] = df['image_text'].values
display(topic_encoded[["text", "topic_1", "topic_2"]])

text   topic_1   topic_2
0    Behaviour Therapist (ABA Therapy) Workng w chi...  3.364208  0.343514
1    MALAYSIA'S MEDIAN SALARY LW 19 bY a bY BELOW T...  0.135947 -0.053438
2    Director 32 Male Chinese 9 years Bachelors 5-1...  2.165237  0.733433
3    IT (Healthcare / Cybersecurity) 45 Malay RM 60...  2.336879 -1.365057
4    PR/Marketing in NGO sector 28 Female Malay 6 y...  1.207181  0.355335
..                                                 ...       ...       ...
151  Lead data science, manager, industry data anal...  1.554612 -0.027796
152                              2 P ~— = rR a. EN a €  0.008018  0.012404
153  Mobile Developer, Fintech MNC 22, East MY Bumi...  1.275746  0.102072
154  strange for someone who's "supposed" to be hol...  2.134513  0.067230
155  QUE IENAIEIN 8y4242)© submission was annoynmou...  2.125097 -0.353518

[156 rows x 3 columns]

In [ ]:
# extract first 2 words of image_text
# text = [' '.join(text.split()[:2]) for text in image_text]
# text[:5]

### Encoding

In [36]:
dictionary = CV.get_feature_names()
dictionary[:10]

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


['000', '10', '100', '10k', '10pm', '12', '12k', '14', '15', '15k']

### The Encoding Matrix

In [37]:
encoding_matrix = pd.DataFrame(svd.components_,
                               index=['topic_1', 'topic_2'],
                               columns=dictionary).T
encoding_matrix

topic_1   topic_2
000      0.031436 -0.024965
10       0.050864  0.015893
100      0.005908 -0.001912
10k      0.007605 -0.000214
10pm     0.002520 -0.001951
...           ...       ...
ya       0.005097 -0.000965
years    0.027826  0.020558
young    0.026325  0.005498
younger  0.006324  0.001477
yrs      0.013707  0.006872

[1022 rows x 2 columns]

### Find the top words for each topic

In [38]:
encoding_matrix['abs_topic_1'] = np.abs(encoding_matrix['topic_1'])
encoding_matrix['abs_topic_2'] = np.abs(encoding_matrix['topic_2'])
top_20_topic1 = encoding_matrix.sort_values('abs_topic_1', ascending=False)
top_20_topic1.head(20)

topic_1   topic_2  abs_topic_1  abs_topic_2
time       0.250149 -0.348853     0.250149     0.348853
veri       0.184425  0.195952     0.184425     0.195952
rm         0.175356 -0.199679     0.175356     0.199679
day        0.145593 -0.093089     0.145593     0.093089
need       0.145433  0.025384     0.145433     0.025384
know       0.144180  0.057089     0.144180     0.057089
onli       0.135349  0.017134     0.135349     0.017134
5k         0.135180 -0.038523     0.135180     0.038523
basic      0.124083 -0.052454     0.124083     0.052454
increment  0.120838 -0.003670     0.120838     0.003670
becaus     0.114331  0.000135     0.114331     0.000135
better     0.113887 -0.084454     0.113887     0.084454
engin      0.113293 -0.031139     0.113293     0.031139
hour       0.112680 -0.065772     0.112680     0.065772
mani       0.109813 -0.000766     0.109813     0.000766
alway      0.105094 -0.081135     0.105094     0.081135
paid       0.103848  0.097153     0.103848     0.097153
company    0.101364 -0.010214     0.101364     0.010214
peopl      0.100553  0.022993     0.100553     0.022993
lot        0.099599 -0.005940     0.099599     0.005940

We can see that most of the top 20 words for **Topic 1** are more close to describing satisfied work, such as "better", "increment.Therefore, **Topic 1** refers to 'satisfied'.

In [39]:
top_20_topic2 = encoding_matrix.sort_values('abs_topic_2', ascending=False)
top_20_topic2.head(20)

topic_1   topic_2  abs_topic_1  abs_topic_2
lawyer     0.087621  0.478234     0.087621     0.478234
time       0.250149 -0.348853     0.250149     0.348853
firm       0.068533  0.232817     0.068533     0.232817
law        0.050257  0.221758     0.050257     0.221758
rm         0.175356 -0.199679     0.175356     0.199679
veri       0.184425  0.195952     0.184425     0.195952
underpaid  0.079781  0.143078     0.079781     0.143078
especi     0.072281  0.140117     0.072281     0.140117
client     0.069323  0.114884     0.069323     0.114884
offer      0.095396 -0.113290     0.095396     0.113290
practic    0.020741  0.107944     0.020741     0.107944
field      0.060368  0.105735     0.060368     0.105735
compar     0.051812  0.098831     0.051812     0.098831
paid       0.103848  0.097153     0.103848     0.097153
day        0.145593 -0.093089     0.145593     0.093089
wage       0.061465 -0.091461     0.061465     0.091461
general    0.027863  0.087660     0.027863     0.087660
region     0.014851  0.085464     0.014851     0.085464
better     0.113887 -0.084454     0.113887     0.084454
bad        0.051413  0.082939     0.051413     0.082939

We can see that most of the top 20 words for **Topic 2** are more close to describing dissatisfied work, such as "underpaid", "bad". Therefore, **Topic 2** refers to **'dissatisfied'**.

# Label topic 

In [44]:
topic_encoded = topic_encoded.rename(columns={'topic_1': 'satisfied', 'topic_2': 'dissatisfied'})
display(topic_encoded[["text", "satisfied", "dissatisfied"]])

text  satisfied  \
0    Behaviour Therapist (ABA Therapy) Workng w chi...   3.364208   
1    MALAYSIA'S MEDIAN SALARY LW 19 bY a bY BELOW T...   0.135947   
2    Director 32 Male Chinese 9 years Bachelors 5-1...   2.165237   
3    IT (Healthcare / Cybersecurity) 45 Malay RM 60...   2.336879   
4    PR/Marketing in NGO sector 28 Female Malay 6 y...   1.207181   
..                                                 ...        ...   
151  Lead data science, manager, industry data anal...   1.554612   
152                              2 P ~— = rR a. EN a €   0.008018   
153  Mobile Developer, Fintech MNC 22, East MY Bumi...   1.275746   
154  strange for someone who's "supposed" to be hol...   2.134513   
155  QUE IENAIEIN 8y4242)© submission was annoynmou...   2.125097   

     dissatisfied  
0        0.343514  
1       -0.053438  
2        0.733433  
3       -1.365057  
4        0.355335  
..            ...  
151     -0.027796  
152      0.012404  
153      0.102072  
154      0.067230  
155     -0.353518  

[156 rows x 3 columns]

## Topic Modelling

In [49]:
topic_encoded["dissatisfied"] = np.abs(topic_encoded["dissatisfied"])
topic_encoded["satisfied"] = np.abs(topic_encoded["satisfied"])

In [50]:
topic_encoded["label"] = topic_encoded.apply(lambda row: 'satisfied' if row["satisfied"] >= row["dissatisfied"] else 'dissatisfied', axis=1)
display(topic_encoded[["text", "satisfied", "dissatisfied", "label"]])

text  satisfied  \
0    Behaviour Therapist (ABA Therapy) Workng w chi...   3.364208   
1    MALAYSIA'S MEDIAN SALARY LW 19 bY a bY BELOW T...   0.135947   
2    Director 32 Male Chinese 9 years Bachelors 5-1...   2.165237   
3    IT (Healthcare / Cybersecurity) 45 Malay RM 60...   2.336879   
4    PR/Marketing in NGO sector 28 Female Malay 6 y...   1.207181   
..                                                 ...        ...   
151  Lead data science, manager, industry data anal...   1.554612   
152                              2 P ~— = rR a. EN a €   0.008018   
153  Mobile Developer, Fintech MNC 22, East MY Bumi...   1.275746   
154  strange for someone who's "supposed" to be hol...   2.134513   
155  QUE IENAIEIN 8y4242)© submission was annoynmou...   2.125097   

     dissatisfied         label  
0        0.343514     satisfied  
1        0.053438     satisfied  
2        0.733433     satisfied  
3        1.365057     satisfied  
4        0.355335     satisfied  
..            ...           ...  
151      0.027796     satisfied  
152      0.012404  dissatisfied  
153      0.102072     satisfied  
154      0.067230     satisfied  
155      0.353518     satisfied  

[156 rows x 4 columns]

In [51]:
topic_encoded['label'].value_counts()

satisfied       152
dissatisfied      4
Name: label, dtype: int64

In [53]:
topic_encoded['text'][152]

'2 P ~— = rR a. EN a €'